system-contracts/test/L1Messenger.spec.ts

```diff
@@ -0,0 +1,517 @@
+  let stateDiffsSetupData: StateDiffSetupData;
+  let logData: LogData;
+  let logs: string[];
+  let bytecodeData: BytecodeData;
+  let emulator: L1MessengerPubdataEmulator;
+
+  before(async () => {
+    await prepareEnvironment();
+    wallet = getWallets()[0];
+    await deployContractOnAddress(TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS, "L1Messenger");
+    l1Messenger = L1MessengerFactory.connect(TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS, wallet);
+    l1MessengerAccount = await ethers.getImpersonatedSigner(TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS);
+    knownCodeStorageAccount = await ethers.getImpersonatedSigner(TEST_KNOWN_CODE_STORAGE_CONTRACT_ADDRESS);
+    bootloaderAccount = await ethers.getImpersonatedSigner(TEST_BOOTLOADER_FORMAL_ADDRESS);
+    // setup
+    stateDiffsSetupData = await setupStateDiffs();
+    logData = setupLogData();
+    logs = createLogs(l1MessengerAccount, l1Messenger, logData);
+    bytecodeData = await setupBytecodeData(l1Messenger.address);
+    await setResult("SystemContext", "txNumberInBlock", [], {
+      failure: false,
+      returnData: ethers.utils.defaultAbiCoder.encode(["uint16"], [1]),
+    });
+    emulator = new L1MessengerPubdataEmulator();
+  });
+
+  // this part is necessary to clean the state of L1Messenger contract
+  after(async () => {
+    emulator.addLog(logs[0]);
+    const fourthLog = ethers.utils.concat([
+      ethers.utils.hexlify([0]),
+      ethers.utils.hexlify([0]),
+      ethers.utils.hexZeroPad(ethers.utils.hexlify(1), 2),
+      ethers.utils.hexZeroPad(l1MessengerAccount.address, 20),
+      logData.key,
+      logData.value,
+    ]);
+    emulator.addLog(fourthLog);
+    emulator.addLog(logs[1]);
+    emulator.addMessage({ lengthBytes: logData.currentMessageLengthBytes, content: logData.message });
+    emulator.addBytecode(bytecodeData);
+    await l1Messenger
+      .connect(bootloaderAccount)
+      .publishPubdataAndClearState(emulator.buildTotalL2ToL1PubdataAndStateDiffs());
+    await network.provider.request({
+      method: "hardhat_stopImpersonatingAccount",
+      params: [TEST_L1_MESSENGER_SYSTEM_CONTRACT_ADDRESS],
+    });
+    await network.provider.request({
+      method: "hardhat_stopImpersonatingAccount",
+      params: [TEST_KNOWN_CODE_STORAGE_CONTRACT_ADDRESS],
+    });
+    await network.provider.request({
+      method: "hardhat_stopImpersonatingAccount",
+      params: [TEST_BOOTLOADER_FORMAL_ADDRESS],
+    });
+  });
+
+  describe("publishPubdataAndClearState", async () => {
+    it("publishPubdataAndClearState passes correctly", async () => {
+      await (
+        await l1Messenger.connect(l1MessengerAccount).sendL2ToL1Log(logData.isService, logData.key, logData.value)
+      ).wait();
+      emulator.addLog(logs[0]);
+      await (await l1Messenger.connect(l1MessengerAccount).sendToL1(logData.message)).wait();
+      emulator.addLog(logs[1]);
+      emulator.addMessage({ lengthBytes: logData.currentMessageLengthBytes, content: logData.message });
```

Comment:
`logs[1]` should be a part of `logData` since these are connected together

Reply:
fixed